mindmeld/models/taggers/pytorch_crf.py

```diff
@@ -0,0 +1,408 @@
+            labels = []
+        if fit:
+            if self.feat_extract_type == "dict":
+                comb_dict_list = [x for seq in feat_dicts for x in seq]
+                self.feat_extractor.fit(comb_dict_list)
+                self.num_feats = len(self.feat_extractor.get_feature_names())
+            if labels:
+                self.label_encoder.fit([x for l in labels for x in l])
+                self.pad_index = len(self.label_encoder.classes_) - 1
+                self.classes = self.label_encoder.classes_
+                self.num_classes = len(self.label_encoder.classes_)
+
+            self.ready = True
+        feats = []
+        encoded_labels = []
+        seq_lens = [len(x) for x in feat_dicts]
+        max_seq_len = max(seq_lens)
+
+        for i, (x, y) in enumerate(zip_longest(feat_dicts, labels)):
+
+            padded_x = x + [{}] * (max_seq_len - seq_lens[i])
+            sparse_feat = self.feat_extractor.transform(padded_x).tocoo()
+            sparse_feat_tensor = torch.sparse_coo_tensor(
+                indices=torch.as_tensor(np.stack([sparse_feat.row, sparse_feat.col])),
+                values=torch.as_tensor(sparse_feat.data), size=sparse_feat.shape)
+            feats.append(sparse_feat_tensor)
+
+            if y:
+                transformed_label = self.label_encoder.transform(y)
+                transformed_label = np.pad(transformed_label, pad_width=(0, max_seq_len - seq_lens[i]),
+                                           constant_values=self.pad_index)
+                label_tensor = torch.as_tensor(transformed_label, dtype=torch.long)
+                encoded_labels.append(label_tensor)
+        return (feats, encoded_labels, seq_lens) if encoded_labels else (feats, seq_lens)
+
+
+# pylint: disable=too-many-instance-attributes
+class TorchCrfModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.optimizer = None
+        self.encoder = None
+        self.best_model_save_path = os.path.join(USER_CONFIG_DIR, "tmp", str(uuid.uuid4()), "best_crf_model.pt")
+        os.makedirs(os.path.dirname(self.best_model_save_path), exist_ok=True)
+
+    def set_random_states(self):
+        torch.manual_seed(self.random_state)
+        random.seed(self.random_state + 1)
+        np.random.seed(self.random_state + 2)
+
+    def validate_params(self):
+        if self.optimizer_type not in ["sgd", "adam"]:
+            raise MindMeldError(
+                f"Optimizer type {self.optimizer_type} not supported. Supported options are ['sgd', 'adam']")
+        elif self.feat_type not in ["hash", "dict"]:
+            raise MindMeldError(f"Feature type {self.feat_type} not supported. Supported options are ['hash', 'dict']")
+        elif not 0 < self.train_val_split < 1:
+            raise MindMeldError("Train-dev split should be a value between 0 and 1.")
+        elif not 0 <= self.drop_input < 1:
+            raise MindMeldError("Drop Input should be a value between 0 and 1. (inclusive)")
```

Comment:
These can just be single `if`'s, since you're raising if they're true.

Reply:
Fixed